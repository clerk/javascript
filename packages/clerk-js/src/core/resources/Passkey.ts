import { ClerkWebAuthnError } from '@clerk/shared/error';
import { isWebAuthnPlatformAuthenticatorSupported, isWebAuthnSupported } from '@clerk/shared/webauthn';
import type {
  DeletedObjectJSON,
  DeletedObjectResource,
  PasskeyJSON,
  PasskeyResource,
  PasskeyVerificationResource,
  PublicKeyCredentialWithAuthenticatorAttestationResponse,
  UpdatePasskeyParams,
} from '@clerk/types';

import { unixEpochToDate } from '../../utils/date';
import { serializePublicKeyCredential, webAuthnCreateCredential } from '../../utils/passkeys';
import { clerkMissingWebAuthnPublicKeyOptions } from '../errors';
import { BaseResource, DeletedObject, PasskeyVerification } from './internal';

export class Passkey extends BaseResource implements PasskeyResource {
  id!: string;
  pathRoot = '/me/passkeys';
  verification: PasskeyVerificationResource | null = null;
  name: string | null = null;
  lastUsedAt: Date | null = null;
  createdAt!: Date;
  updatedAt!: Date;

  public constructor(data: PasskeyJSON) {
    super();
    this.fromJSON(data);
  }

  private static async create() {
    return BaseResource._fetch({
      path: `/me/passkeys`,
      method: 'POST',
    }).then(res => new Passkey(res?.response as PasskeyJSON));
  }

  private static async attemptVerification(
    passkeyId: string,
    credential: PublicKeyCredentialWithAuthenticatorAttestationResponse,
  ) {
    const jsonPublicKeyCredential = serializePublicKeyCredential(credential);
    return BaseResource._fetch({
      path: `/me/passkeys/${passkeyId}/attempt_verification`,
      method: 'POST',
      body: { strategy: 'passkey', publicKeyCredential: JSON.stringify(jsonPublicKeyCredential) } as any,
    }).then(res => new Passkey(res?.response as PasskeyJSON));
  }

  /**
   * Developers should not be able to create a new Passkeys from an already instanced object
   */
  static async registerPasskey() {
    /**
     * The UI should always prevent from this method being called if WebAuthn is not supported.
     * As a precaution we need to check if WebAuthn is supported.
     */
    const _isWebAuthnSupported = Passkey.clerk.__unstable__isWebAuthnSupported || isWebAuthnSupported;
    const _webAuthnCreateCredential = Passkey.clerk.__unstable__createPublicCredentials || webAuthnCreateCredential;
    const _isWebAuthnPlatformAuthenticatorSupported =
      Passkey.clerk.__unstable__isWebAuthnPlatformAuthenticatorSupported || isWebAuthnPlatformAuthenticatorSupported;

    if (!_isWebAuthnSupported()) {
      throw new ClerkWebAuthnError('Passkeys are not supported on this device.', {
        code: 'passkey_not_supported',
      });
    }

    const passkey = await this.create();

    const { verification } = passkey;

    const publicKey = verification?.publicKey;

    // This should never occur, just a fail-safe
    if (!publicKey) {
      clerkMissingWebAuthnPublicKeyOptions('create');
    }

    if (publicKey.authenticatorSelection?.authenticatorAttachment === 'platform') {
      if (!(await _isWebAuthnPlatformAuthenticatorSupported())) {
        throw new ClerkWebAuthnError(
          'Registration requires a platform authenticator but the device does not support it.',
          {
            code: 'passkey_pa_not_supported',
          },
        );
      }
    }

    // Invoke the WebAuthn create() method.
    const { publicKeyCredential, error } = await _webAuthnCreateCredential(publicKey);

    if (!publicKeyCredential) {
      throw error;
    }
    return this.attemptVerification(passkey.id, publicKeyCredential);
  }

  /**
   * PATCH /v1/me/passkeys/{passkeyIdentificationID}
   */
  update = (params: UpdatePasskeyParams): Promise<PasskeyResource> =>
    this._basePatch({
      body: params,
    });

  /**
   * DELETE /v1/me/passkeys/{passkeyIdentificationID}
   */
  delete = async (): Promise<DeletedObjectResource> => {
    const json = (
      await BaseResource._fetch<DeletedObjectJSON>({
        path: this.path(),
        method: 'DELETE',
      })
    )?.response as unknown as DeletedObjectJSON;

    return new DeletedObject(json);
  };

  protected fromJSON(data: PasskeyJSON | null): this {
    if (!data) {
      return this;
    }

    this.id = data.id;
    this.name = data.name;
    this.lastUsedAt = data.last_used_at ? unixEpochToDate(data.last_used_at) : null;
    this.createdAt = unixEpochToDate(data.created_at);
    this.updatedAt = unixEpochToDate(data.updated_at);

    if (data.verification) {
      this.verification = new PasskeyVerification(data.verification);
    }
    return this;
  }
}
