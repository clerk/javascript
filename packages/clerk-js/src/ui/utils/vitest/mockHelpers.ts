// import { jest } from '@jest/globals';
import type { LoadedClerk } from '@clerk/types';
import type { ActiveSessionResource } from '@clerk/types';
import { type Mocked, vi } from 'vitest';

import type { RouteContextValue } from '../../router';

type FunctionLike = (...args: any) => any;

type DeepVitestMocked<T> = T extends FunctionLike
  ? Mocked<T>
  : T extends object
    ? {
        [k in keyof T]: DeepVitestMocked<T[k]>;
      }
    : T;

// Removing jest.Mock type for now, relying on inference
type MockMap<T = any> = {
  [K in { [K in keyof T]: T[K] extends (...args: any[]) => any ? K : never }[keyof T]]?: ReturnType<typeof vi.fn>;
};

const mockProp = <T>(obj: T, k: keyof T, mocks?: MockMap<T>) => {
  if (typeof obj[k] === 'function') {
    const mockFn = mocks?.[k] ?? vi.fn();
    (obj[k] as unknown as ReturnType<typeof vi.fn>) = mockFn;
  }
};

const mockMethodsOf = <T extends Record<string, any> | null = any>(
  obj: T,
  options?: {
    exclude: (keyof T)[];
    mocks: MockMap<T>;
  },
) => {
  if (!obj) {
    return;
  }
  Object.keys(obj)
    .filter(key => !options?.exclude.includes(key as keyof T))
    // Pass the specific MockMap for the object T being mocked
    .forEach(k => mockProp(obj, k, options?.mocks));
};

export const mockClerkMethods = (clerk: LoadedClerk): DeepVitestMocked<LoadedClerk> => {
  // Cast clerk to any to allow mocking properties
  const clerkAny = clerk as any;

  mockMethodsOf(clerkAny);
  if (clerkAny.client) {
    mockMethodsOf(clerkAny.client.signIn);
    mockMethodsOf(clerkAny.client.signUp);
    clerkAny.client.sessions?.forEach((session: ActiveSessionResource) => {
      const sessionAny = session as any;
      mockMethodsOf(sessionAny, {
        exclude: ['checkAuthorization'],
        mocks: {
          // Ensure touch mock matches expected signature if available, otherwise basic mock
          touch: vi.fn(() => Promise.resolve(session)),
        },
      });
      if (sessionAny.user) {
        mockMethodsOf(sessionAny.user);
        sessionAny.user.emailAddresses?.forEach((m: any) => mockMethodsOf(m));
        sessionAny.user.phoneNumbers?.forEach((m: any) => mockMethodsOf(m));
        sessionAny.user.externalAccounts?.forEach((m: any) => mockMethodsOf(m));
        sessionAny.user.organizationMemberships?.forEach((m: any) => {
          mockMethodsOf(m);
          if (m.organization) {
            mockMethodsOf(m.organization);
          }
        });
        sessionAny.user.passkeys?.forEach((m: any) => mockMethodsOf(m));
      }
    });
  }
  mockProp(clerkAny, 'navigate');
  mockProp(clerkAny, 'setActive');
  mockProp(clerkAny, 'redirectWithAuth');
  mockProp(clerkAny, '__internal_navigateWithError');
  return clerkAny as DeepVitestMocked<LoadedClerk>;
};

export const mockRouteContextValue = ({ queryString = '' }: Partial<DeepVitestMocked<RouteContextValue>>) => {
  return {
    basePath: '',
    startPath: '',
    flowStartPath: '',
    fullPath: '',
    indexPath: '',
    currentPath: '',
    queryString,
    queryParams: {},
    getMatchData: vi.fn(),
    matches: vi.fn(),
    baseNavigate: vi.fn(),
    navigate: vi.fn(() => Promise.resolve(true)),
    resolve: vi.fn((to: string) => new URL(to, 'https://clerk.com')),
    refresh: vi.fn(),
    params: {},
  } as RouteContextValue; // Keep original type assertion, DeepVitestMocked applied to input only
};
