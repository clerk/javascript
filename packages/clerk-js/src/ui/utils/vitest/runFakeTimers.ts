import { vi } from 'vitest';
// import { jest } from '@jest/globals';
import { act } from '@testing-library/react';

type WithAct = <T>(fn: T) => T;
const withAct = ((fn: any) =>
  (...args: any) => {
    act(() => {
      fn(...args);
    });
  }) as WithAct;

const advanceTimersByTime = withAct(vi.advanceTimersByTime.bind(vi));
const runAllTimers = withAct(vi.runAllTimers.bind(vi));
const runOnlyPendingTimers = withAct(vi.runOnlyPendingTimers.bind(vi));

const createFakeTimersHelpers = () => {
  return { advanceTimersByTime, runAllTimers, runOnlyPendingTimers };
};

type FakeTimersHelpers = ReturnType<typeof createFakeTimersHelpers>;
type RunFakeTimersCallback = (timers: FakeTimersHelpers) => void | Promise<void>;

export const runFakeTimers = <T extends RunFakeTimersCallback, R extends ReturnType<T>>(
  cb: T,
): R extends Promise<any> ? Promise<void> : void => {
  vi.useFakeTimers();
  const res = cb(createFakeTimersHelpers());
  if (res && 'then' in res) {
    // @ts-expect-error
    return res.finally(() => vi.useRealTimers());
  }
  vi.useRealTimers();
  // @ts-ignore
  return;
};
