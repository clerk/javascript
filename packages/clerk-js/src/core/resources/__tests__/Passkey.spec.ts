import { describe, expect, it } from 'vitest';

import { Passkey } from '../Passkey';

describe('Passkey', () => {
  it('has the same initial properties', () => {
    const passkey = new Passkey({
      object: 'passkey',
      id: 'passkey_123',
      name: 'My Device',
      credential_id: 'credential_abc123',
      verification: {
        object: 'verification',
        id: 'verification_123',
        status: 'verified',
        strategy: 'passkey',
        attempts: 1,
        expire_at: 1735689700000,
        verified_at_client: null,
        error: null,
      },
      last_used_at: 1735689600000,
      created_at: 1735689500000,
      updated_at: 1735689650000,
    });

    expect(passkey).toMatchObject({
      id: 'passkey_123',
      name: 'My Device',
      credentialId: 'credential_abc123',
      verification: expect.any(Object),
      lastUsedAt: expect.any(Date),
      createdAt: expect.any(Date),
      updatedAt: expect.any(Date),
    });
  });
});

describe('Passkey Snapshots', () => {
  it('should match snapshot for passkey structure', () => {
    const passkey = new Passkey({
      object: 'passkey',
      id: 'passkey_123',
      name: 'MacBook Pro Touch ID',
      credential_id: 'cred_abcd1234efgh5678',
      verification: {
        object: 'verification',
        id: 'verification_123',
        status: 'verified',
        strategy: 'passkey',
        attempts: 1,
        expire_at: 1735689700000,
        verified_at_client: null,
        error: null,
      },
      last_used_at: 1735689600000,
      created_at: 1735689500000,
      updated_at: 1735689650000,
    });

    const snapshot = {
      id: passkey.id,
      name: passkey.name,
      credentialId: passkey.credentialId,
      verification: {
        id: passkey.verification?.id,
        status: passkey.verification?.status,
        strategy: passkey.verification?.strategy,
      },
      lastUsedAt: passkey.lastUsedAt?.getTime(),
      createdAt: passkey.createdAt?.getTime(),
      updatedAt: passkey.updatedAt?.getTime(),
    };

    expect(snapshot).toMatchSnapshot();
  });

  it('should match snapshot for __internal_toSnapshot method', () => {
    const passkey = new Passkey({
      object: 'passkey',
      id: 'passkey_456',
      name: 'iPhone Face ID',
      credential_id: 'cred_xyz789abc012',
      verification: {
        object: 'verification',
        id: 'verification_456',
        status: 'unverified',
        strategy: 'passkey',
        attempts: 0,
        expire_at: 1735689800000,
        verified_at_client: null,
        error: null,
      },
      last_used_at: null,
      created_at: 1735689500000,
      updated_at: 1735689500000,
    });

    if (typeof passkey.__internal_toSnapshot === 'function') {
      const snapshot = passkey.__internal_toSnapshot();
      expect(snapshot).toMatchSnapshot();
    }
  });
});
