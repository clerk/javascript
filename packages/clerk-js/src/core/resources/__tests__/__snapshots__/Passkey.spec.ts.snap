// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Passkey Snapshots > should match snapshot for __internal_toSnapshot method 1`] = `
{
  "created_at": 1735689500000,
  "id": "passkey_456",
  "last_used_at": null,
  "name": "iPhone Face ID",
  "object": "passkey",
  "updated_at": 1735689500000,
  "verification": {
    "attempts": 0,
    "error": {
      "code": "",
      "long_message": undefined,
      "message": "",
      "meta": {
        "email_addresses": undefined,
        "identifiers": undefined,
        "param_name": undefined,
        "plan": undefined,
        "session_id": undefined,
        "zxcvbn": undefined,
      },
    },
    "expire_at": 1735689800000,
    "external_verification_redirect_url": null,
    "id": "",
    "message": null,
    "nonce": null,
    "object": "verification",
    "status": "unverified",
    "strategy": "passkey",
    "verified_at_client": null,
  },
}
`;

exports[`Passkey Snapshots > should match snapshot for passkey structure 1`] = `
{
  "createdAt": 1735689500000,
  "credentialId": undefined,
  "id": "passkey_123",
  "lastUsedAt": 1735689600000,
  "name": "MacBook Pro Touch ID",
  "updatedAt": 1735689650000,
  "verification": {
    "id": undefined,
    "status": "verified",
    "strategy": "passkey",
  },
}
`;
